File: src/TaskOTime/TaskOTime.TimeTrackingServices/ObservableObject.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Runtime.CompilerServices

Namespace ActiveDevelop.TimeTrackingServices
    Public MustInherit Class ObservableObject
        Implements INotifyPropertyChanged

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        ''' <summary>
        '''  vergelijkt met de standaardgelijkheid en meldt alleen wanneer de opgeslagen waarde verandert.
        ''' </summary>
        ''' <returns>
        '''  waar nadat opslag en melding zijn uitgevoerd; onwaar wanneer beide waarden gelijk zijn.
        ''' </returns>
        ''' <remarks>
        '''  de opslag gaat aan de melding vooraf.  een uitzondering uit een afnemer wordt doorgegeven,
        '''  zonder dat deze helper de opgeslagen waarde terugzet.
        ''' </remarks>
        Protected Function SetProperty(Of T)(ByRef storage As T, value As T, <CallerMemberName> Optional propertyName As String = Nothing) As Boolean
            If EqualityComparer(Of T).Default.Equals(storage, value) Then
                Return False
            End If

            storage = value
            OnPropertyChanged(propertyName)
            Return True
        End Function

        ''' <summary>
        '''  geeft de eigenschapsnaam direct door, zonder zelf gelijkheid of afhankelijkheden te toetsen.
        ''' </summary>
        Protected Overridable Sub OnPropertyChanged(<CallerMemberName> Optional propertyName As String = Nothing)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub

        Protected Overridable Sub OnPropertyChanged(args As PropertyChangedEventArgs)
            RaiseEvent PropertyChanged(Me, args)
        End Sub
    End Class
End Namespace

File: src/TaskOTime/TaskOTime.TimeTrackingServices/ObservableObject.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace ActiveDevelop.TimeTrackingServices
{
    public abstract class ObservableObject : INotifyPropertyChanged
    {

        public event PropertyChangedEventHandler PropertyChanged;

        /// <summary>
        /// Uses the default equality comparer and notifies subscribers only when the stored value changes.
        /// </summary>
        /// <returns>
        /// True after storing the value and completing the notification; false when the values are equal.
        /// </returns>
        /// <remarks>
        /// Storage is updated before notification. Subscriber exceptions propagate without this helper
        /// restoring the previous value.
        /// </remarks>
        protected bool SetProperty<T>(ref T storage, T value, [CallerMemberName] string propertyName = null)
        {
            if (EqualityComparer<T>.Default.Equals(storage, value))
            {
                return false;
            }

            storage = value;
            OnPropertyChanged(propertyName);
            return true;
        }

        /// <summary>
        /// Forwards the property name directly, without checking equality or dependent properties.
        /// </summary>
        protected virtual void OnPropertyChanged([CallerMemberName] string propertyName = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        protected virtual void OnPropertyChanged(PropertyChangedEventArgs args)
        {
            PropertyChanged?.Invoke(this, args);
        }
    }
}